src/submission/merge-submission.ts

```diff
@@ -0,0 +1,105 @@
+import { DeepReadonly } from 'deep-freeze';
+import { Donor, Specimen, Sample } from '../clinical/clinical-entities';
+import { ActiveClinicalSubmission } from './submission-entities';
+import { FileType } from './submission-api';
+import _ from 'lodash';
+import { loggerFor } from '../logger';
+import { update } from './schema-api';
+const L = loggerFor(__filename);
+
+type ClinicalEnitityRecord = DeepReadonly<
+  Readonly<{
+    [key: string]: string;
+  }>
+>;
+
+export const mergeActiveSubmissionWithDonors = async (
+  activeSubmission: DeepReadonly<ActiveClinicalSubmission>,
+  donors: readonly DeepReadonly<Donor>[],
+) => {
+  const updatedDonors = _.cloneDeep(donors as Donor[]);
+  for (const entityType in activeSubmission.clinicalEntities) {
+    const entityData = activeSubmission.clinicalEntities[entityType];
+
+    // Find the donor that matches each record, and update the entity within that donor
+    entityData.records.forEach(record => {
+      const donorId = record.submitter_donor_id;
+      const donor = _.find(updatedDonors, ['submitterId', donorId]);
+      if (donor) {
+        switch (entityType) {
+          case FileType.DONOR:
+            updateDonorFromSubmissionRecord(donor, record);
+            break;
+          case FileType.SPECIMEN:
+            updateSpecimenRecord(donor, record);
+            break;
+          case FileType.SAMPLE:
+            updateSampleRecord(donor, record);
+            break;
+          default:
+            addOrUpdateClinicalRecord(donor, record, entityType);
+            break;
+        }
+      }
+    });
+  }
+
+  return updatedDonors;
+};
+
+const updateDonorFromSubmissionRecord = (donor: Donor, record: ClinicalEnitityRecord) => {
+  donor.gender = record.gender;
+  donor.clinicalInfo = _.omit(record, ['program_id', 'submitter_donor_id', 'gender']);
+};
+const updateSpecimenRecord = (donor: Donor, record: ClinicalEnitityRecord) => {
+  // Find specimen in donor
+  const specimen = _.find(donor.specimens, [
```

Comment:
what happens if the specimen doesn't exist?

Reply:
If the specimen is not there we should break but in a controlled way. I'll add some checking and throw some intentional errors.

Reply:
Same with specimen.